secureboot_policy.go

```diff
@@ -1165,117 +1193,136 @@ func (g *secureBootPolicyGen) processOSLoadEvent(paths []*secureBootPolicyGenPat
-		return nil, xerrors.Errorf("cannot identify initial OS launch verification event: %w", err)
+		return xerrors.Errorf("cannot identify initial OS launch verification event: %w", err)
 	}
 
-	var allPaths []*secureBootPolicyGenPath
-
+	// Process the pre-OS events for the current signature DB and then with each pending update applied
+	// in turn.
+	var roots []*secureBootPolicyGenBranch
 	for i := 0; i <= len(sigDbUpdates); i++ {
-		path := &secureBootPolicyGenPath{pcrValue: make(tpm2.Digest, g.PCRAlgorithm.Size()), dbUpdateLevel: i}
-		if err := g.processPreOSEvents(path, events, initialOSVerificationEvent, sigDbUpdates[0:i]); err != nil {
-			return nil, xerrors.Errorf("cannot process pre-OS events from event log: %w", err)
+		branch := &secureBootPolicyGenBranch{gen: g, profile: NewPCRProtectionProfile(), dbUpdateLevel: i}
+		if err := branch.processPreOSEvents(events, initialOSVerificationEvent, sigDbUpdates[0:i]); err != nil {
+			return xerrors.Errorf("cannot process pre-OS events from event log: %w", err)
 		}
-		allPaths = append(allPaths, path)
+		roots = append(roots, branch)
 	}
 
-	numPreOSPaths := len(allPaths)
+	allBranches := make([]*secureBootPolicyGenBranch, len(roots))
+	copy(allBranches, roots)
 
-	duplicatePaths := func(paths []*secureBootPolicyGenPath) (out []*secureBootPolicyGenPath) {
-		for _, p := range paths {
-			p2 := p.duplicate()
-			allPaths = append(allPaths, p2)
-			out = append(out, p2)
-		}
-		return
-	}
-
-	var loadEvents []*loadEventAndPaths
-	var nextLoadEvents []*loadEventAndPaths
+	var loadEvents []*sbLoadEventAndBranches
+	var nextLoadEvents []*sbLoadEventAndBranches
 
-	for i, e := range g.LoadSequences {
-		var paths []*secureBootPolicyGenPath
-		if i == 0 {
-			paths = allPaths
-		} else {
-			paths = duplicatePaths(allPaths[0:numPreOSPaths])
+	if len(g.LoadSequences) == 1 {
+		loadEvents = append(loadEvents, &sbLoadEventAndBranches{event: g.LoadSequences[0], branches: roots})
+	} else {
+		for _, e := range g.LoadSequences {
+			var branches []*secureBootPolicyGenBranch
+			for _, b := range roots {
+				branches = append(branches, b.branch())
+			}
+			allBranches = append(allBranches, branches...)
+			loadEvents = append(loadEvents, &sbLoadEventAndBranches{event: e, branches: branches})
 		}
-		loadEvents = append(loadEvents, &loadEventAndPaths{event: e, paths: paths})
 	}
 
 	for len(loadEvents) > 0 {
 		e := loadEvents[0]
 		loadEvents = loadEvents[1:]
-		if err := g.processOSLoadEvent(e.paths, e.event); err != nil {
-			return nil, xerrors.Errorf("cannot process OS load event for %s: %w", e.event.Image, err)
+
+		if err := g.processOSLoadEvent(e.branches, e.event); err != nil {
+			return xerrors.Errorf("cannot process OS load event for %s: %w", e.event.Image, err)
 		}
-		for i, n := range e.event.Next {
-			var paths []*secureBootPolicyGenPath
-			if i == 0 {
-				paths = e.paths
-			} else {
-				paths = duplicatePaths(e.paths)
+
+		if len(e.event.Next) == 1 {
+			e.event = e.event.Next[0]
```

Comment:
is this mutation in place required or could we replace loadEvents[0] whole instead?